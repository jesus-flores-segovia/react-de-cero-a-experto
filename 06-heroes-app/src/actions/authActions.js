import { types } from "../types/types";

export const login = (userName) => {
    return ( 
        {
            type: types.login,
            payload: userName
        }
    );
};