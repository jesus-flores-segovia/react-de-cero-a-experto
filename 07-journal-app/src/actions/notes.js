import { addDoc, collection, deleteDoc, doc, updateDoc } from "firebase/firestore/lite";
import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNote = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;
        
        const newNote = {
            title: "",
            body: "",
            date: new Date().getTime()
        }

        const doc = await addDoc(collection(db, `${ uid }`, "journal/notes"),{
            title: '',
            body: '',
            date: new Date().getTime()
        });

        dispatch(setActiveNote(doc.id, newNote));
        dispatch(addNewNote(doc.id, newNote));
    }
};

export const setActiveNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
});

export const setNotes = (notes) => ({
    type: types.notesLoad,
    payload: [...notes]
});

export const startLoadingNotes = () => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;
        
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
};

export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const uid = getState().auth.uid;

        if(!note.imageUrl) {
            delete note.imageUrl;
        }

        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);

        await updateDoc(noteRef, noteToFirestore);

        dispatch(refreshNote(note.id, noteToFirestore));
        Swal.fire("Saved", note.title, "success");
    }
};

export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
});

export const startUploading = (file) => {
    return async (dispatch, getState) => {
        
        const {active} = getState().notes;

        Swal.fire({
            title: "Uploading...",
            text: "Please wait...",
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        })

        const fileUrl = await fileUpload(file);
        active.imageUrl = fileUrl;  
        
        dispatch(startSaveNote(active));
    }
};

export const startDelete = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        Swal.fire({
            title: 'Are you sure?',
            text: "You won't be able to revert this!",
            icon: 'warning',
            showCancelButton: true,
            confirmButtonColor: '#3085d6',
            cancelButtonColor: '#d33',
            confirmButtonText: 'Yes, delete it!'
          }).then(async (result) => {
            if (result.isConfirmed) {
              Swal.fire(
                'Deleted!',
                'Your note has been deleted.',
                'success'
              );

              const noteRef = await doc(db, `${uid}/journal/notes/${id}`);
              await deleteDoc(noteRef);
              dispatch(deleteNote(id));
            }
          })
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: {
        id
    }
})

export const cleanNotes = () => ({
    type: types.notesLogoutCleaning
})