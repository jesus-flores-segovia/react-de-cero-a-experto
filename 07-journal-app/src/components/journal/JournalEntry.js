import React from 'react'

export const JournalEntry = () => {
    return (
        <div className="journal__entry pointer">
            <div 
                className="journal__entry-picture"
                style={{
                    backgroundSize:"cover",
                    backgroundImage: "url(https://whatsup.es/wp-content/uploads/2020/09/travel-trip-journey.jpg)"
                }}
                >
            </div>
            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    A new day
                </p>
                <p className="journal__entry-content">
                    Lorem ipsum asit met
                </p>
            </div>
            <div className="journal__entry-date-box">
                <span>Monday</span>
                <h4>28</h4>
            </div>
        </div>
    )
}
