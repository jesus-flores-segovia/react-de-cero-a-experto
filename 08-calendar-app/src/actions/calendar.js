import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { parseEvents } from "../helpers/parseEvents";
import { types } from "../types/types";

export const startCalendarAddNewEvent = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;
    try {
      const response = await fetchWithToken("events", "POST", event);
      const body = await response.json();

      if (body.ok) {
        event.id = body.event.id;
        event.user = {
          _id: uid,
          name: name,
        };
        dispatch(calendarAddNewEvent(event));
      }
    } catch (error) {
      Swal.fire("Error", "An error has ocurred", "error");
    }
  };
};

export const startCalendarLoadEvents = () => {
  return async (dispatch) => {
    try {
      const response = await fetchWithToken("events", "GET");
      const body = await response.json();

      if (body.ok) {
        dispatch(calendarSetLoadedEvents(parseEvents(body.events)));
      }
    } catch (error) {
      Swal.fire("Error", "An error has ocurred", "error");
    }
  };
};

export const startCalendarUpdateEvent = (event) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithToken(`events/${event.id}`, "PUT", event);
      const body = await response.json();

      if (body.ok) {
        dispatch(calendarUpdateEvent(event));
      } else {
        Swal.fire("Error", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has ocurred", "error");
    }
  };
};

export const startCalendarDeleteEvent = (event) => {
  return async (dispatch) => {
    try {
      Swal.fire({
        title: "Are you sure?",
        html: `Do you want to remove the event '${event.title}'`,
        showDenyButton: true,
        confirmButtonText: "Delete",
        denyButtonText: `Don't Delete`,
      }).then(async (result) => {
        if (result.isConfirmed) {
          const response = await fetchWithToken(
            `events/${event.id}`,
            "DELETE",
            event
          );
          const body = await response.json();

          if (body.ok) {
            Swal.fire(
              "Event deleted!",
              `The event '${event.title}' was deleted succesfully.`,
              "success"
            );
            dispatch(calendarDeleteEvent(event));
          } else {
            Swal.fire("Error", body.msg, "error");
          }
        }
      });
    } catch (error) {
      Swal.fire("Error", "An error has ocurred", "error");
    }
  };
};

const calendarAddNewEvent = (event) => ({
  type: types.calendarNewEvent,
  payload: event,
});

export const calendarSetActiveEvent = (event) => ({
  type: types.calendarSetActiveEvent,
  payload: event,
});

const calendarUpdateEvent = (event) => ({
  type: types.calendarUpdateEvent,
  payload: event,
});

const calendarDeleteEvent = (event) => ({
  type: types.calendarDeleteEvent,
  payload: event,
});

const calendarSetLoadedEvents = (events) => ({
  type: types.calendarSetLoadedEvents,
  payload: events,
});

export const calendarLogoutCleanEvents = () => ({
  type: types.calendarLogoutCleanEvents,
});
