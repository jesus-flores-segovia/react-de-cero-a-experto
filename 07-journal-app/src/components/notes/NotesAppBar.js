import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {

    const {active: note} = useSelector(state => state.notes);
    const dispatch = useDispatch();

    const handleSave = () => {
        dispatch(startSaveNote(note));
    }

    const handlePictureUpload = () => {
        document.querySelector("#fileSelector").click();
    }

    const handleFileChange = (e) => {
        const file = e.target.files[0];

        if(file) {
            dispatch(startUploading(file));
        }
    }

    return (
        <div className="notes__appbar">
            <span>28th August 2021</span>
            <input
                id='fileSelector' 
                type="file" 
                style={{display: 'none'}}
                onChange={handleFileChange}    
                />
            <div>
                <button 
                    className="buttons__btn"
                    onClick={handlePictureUpload}>
                    Picture
                </button>
                <button 
                    className="buttons__btn"
                    onClick={handleSave}
                    >
                    Save
                </button>
            </div>
        </div>
    )
}
