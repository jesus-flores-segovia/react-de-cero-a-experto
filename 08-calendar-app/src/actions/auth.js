import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import { calendarLogoutCleanEvents } from "./calendar";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth",
        { email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { email, password } = body.errors;
          const message = email ? email.msg : password.msg;
          Swal.fire("Cannot login", message, "error");
          return;
        }

        Swal.fire("Cannot login", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    try {
      const response = await fetchWithoutToken(
        "auth/register",
        { name, email, password },
        "POST"
      );
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      } else {
        if (body.errors) {
          const { name, email, password } = body.errors;
          const message = name ? name.msg : email ? email.msg : password.msg;
          Swal.fire("Cannot register", message, "error");
          return;
        }

        Swal.fire("Cannot register", body.msg, "error");
      }
    } catch (error) {
      Swal.fire("Error", "An error has occurred", "error");
    }
  };
};

export const startCheckingLoginState = () => {
  return async (dispatch) => {
    try {
      dispatch(checkLoginState({ isCheckingLoginState: true }));
      const response = await fetchWithToken("auth/newToken");
      const body = await response.json();

      if (body.ok) {
        localStorage.setItem("token", body.token);
        localStorage.setItem("token-init-date", new Date().getTime());
        dispatch(
          login({
            uid: body.uid,
            name: body.name,
          })
        );
      }
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    } catch (error) {
      dispatch(checkLoginState({ isCheckingLoginState: false }));
    }
  };
};

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(calendarLogoutCleanEvents());
    dispatch(logout());
  };
};

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

const checkLoginState = (isCheckingLoginState) => ({
  type: types.authCheckLoginState,
  payload: isCheckingLoginState,
});

const logout = () => ({
  type: types.authLogout,
});
