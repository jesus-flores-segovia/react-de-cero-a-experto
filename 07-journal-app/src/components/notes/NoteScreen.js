import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { setActiveNote, startDelete } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const {active: note} = useSelector(state => state.notes);
    const [values, handleInputChange, reset] = useForm(note);
    const {title, body} = values;

    const activeId = useRef(note.id);

    useEffect(() => {
        if(note.id !== activeId.current) {
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset])

    useEffect(() => {
        dispatch(setActiveNote(values.id, {imageUrl: note.imageUrl, ...values}));
    }, [values, dispatch])

    const handleDelete = () => {
        dispatch(startDelete(note.id));
    }
    
    return (
        <div className="notes__main-content">
            <NotesAppBar/>
            <div className="notes__content">
                <input 
                    name='title'
                    type="text"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    value={title}
                    onChange={handleInputChange}
                />
                <textarea 
                    name='body'
                    placeholder="What happened today"
                    className="notes__textarea" 
                    value={body}
                    onChange={handleInputChange}    
                />
                {note.imageUrl &&
                    <div className="notes__image">
                        <img src={note.imageUrl} alt="image"/>
                    </div>
                }
            </div>
            <button 
                className='buttons__btn buttons__btn-danger'
                onClick={handleDelete}
                >
                Delete
            </button>
        </div>
    )
}
