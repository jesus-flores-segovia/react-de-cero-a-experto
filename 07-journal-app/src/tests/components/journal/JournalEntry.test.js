import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";

import { JournalEntry } from "../../../components/journal/JournalEntry";
import { setActiveNote } from "../../../actions/notes";

const mockStore = configureStore();

const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
  id: 1234,
  title: "Title",
  body: "body",
  date: 0,
  imageUrl: "https://www.test.com",
};

const wrapper = mount(
  <Provider store={store}>
    <JournalEntry {...note} />
  </Provider>
);

describe("Tests inside file 'JournalEntry.test.js'", () => {
  test("Must be shown correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Must call 'setActiveNote' action", () => {
    wrapper.find(".journal__entry").simulate("click");

    expect(store.dispatch).toHaveBeenCalledWith(setActiveNote(note.id, note));
  });
});
