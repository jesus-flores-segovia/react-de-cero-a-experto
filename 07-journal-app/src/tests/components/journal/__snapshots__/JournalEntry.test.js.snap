// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests inside file 'JournalEntry.test.js' Must be shown correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://www.test.com)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Title
    </p>
    <p
      className="journal__entry-content"
    >
      body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Thursday
    </span>
    <h4>
      1st
    </h4>
  </div>
</div>
`;
